xcm/pallet-xcm/src/lib.rs

```diff
@@ -448,9 +638,268 @@ pub mod pallet {
+			origin: OriginFor<T>,
+			location: Box<VersionedMultiLocation>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let location = (*location).try_into().map_err(|()| Error::<T>::BadLocation)?;
+			Self::unrequest_version_notify(location).map_err(|e| {
+				match e {
+					XcmError::InvalidLocation => Error::<T>::NoSubscription,
+					_ => Error::<T>::InvalidOrigin,
+				}
+				.into()
+			})
+		}
 	}
 
 	impl<T: Config> Pallet<T> {
+		/// Will always make progress, and will do its best not to use much more than `weight_cutoff`
+		/// in doing so.
+		pub(crate) fn check_xcm_version_change(
+			mut stage: VersionMigrationStage,
+			weight_cutoff: Weight,
+		) -> (Weight, Option<VersionMigrationStage>) {
+			let mut weight_used = 0;
+
+			// TODO: Correct weights for the components of this:
+			let todo_sv_migrate_weight: Weight = T::DbWeight::get().read + T::DbWeight::get().write;
+			let todo_vn_migrate_weight: Weight = T::DbWeight::get().read + T::DbWeight::get().write;
+			let todo_vnt_already_notified_weight: Weight = T::DbWeight::get().read;
+			let todo_vnt_notify_weight: Weight =
+				T::DbWeight::get().read + T::DbWeight::get().write * 3;
+			let todo_vnt_migrate_weight: Weight =
+				T::DbWeight::get().read + T::DbWeight::get().write;
+			let todo_vnt_migrate_fail_weight: Weight =
+				T::DbWeight::get().read + T::DbWeight::get().write;
+			let todo_vnt_notify_migrate_weight: Weight =
+				T::DbWeight::get().read + T::DbWeight::get().write * 3;
+
+			use VersionMigrationStage::*;
+
+			if stage == MigratateSupportedVersion {
+				// We assume that supported XCM version only ever increases, so just cycle through lower
+				// XCM versioned from the current.
+				for v in 0..XCM_VERSION {
+					for (old_key, value) in SupportedVersion::<T>::drain_prefix(v) {
+						if let Ok(new_key) = old_key.into_latest() {
+							SupportedVersion::<T>::insert(XCM_VERSION, new_key, value);
+						}
+						weight_used.saturating_accrue(todo_sv_migrate_weight);
+						if weight_used >= weight_cutoff {
+							return (weight_used, Some(stage))
+						}
+					}
+				}
+				stage = MigratateVersionNotifiers;
+			}
+			if stage == MigratateVersionNotifiers {
+				for v in 0..XCM_VERSION {
+					for (old_key, value) in VersionNotifiers::<T>::drain_prefix(v) {
+						if let Ok(new_key) = old_key.into_latest() {
+							VersionNotifiers::<T>::insert(XCM_VERSION, new_key, value);
+						}
+						weight_used.saturating_accrue(todo_vn_migrate_weight);
+						if weight_used >= weight_cutoff {
+							return (weight_used, Some(stage))
+						}
+					}
+				}
+				stage = NotifyCurrentTargets;
+			}
+
+			let xcm_version = T::AdvertisedXcmVersion::get();
+			let response = Response::Version(xcm_version);
+
+			if stage == NotifyCurrentTargets {
+				for (key, mut value) in VersionNotifyTargets::<T>::iter_prefix(XCM_VERSION) {
+					let new_key: MultiLocation =
+						match (value.2 == xcm_version, key.clone().try_into()) {
+							(false, Ok(k)) => k,
```

Comment:
I prefer:

```suggestion
						match key.clone().try_into() {
							Ok(k) if value.2 != xcm_version => k,
```

But your call, not going to insist on this.